### PR TITLE
PEP 591: Adding a final qualifier to typing

### PR DESCRIPTION
**Pull request by msullivan**
Initial draft of proposal to add a final qualifier to the typing module

### CODE REVIEW

File: pep-0591.rst [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+will not be assigned to. This is a useful feature in several
+situations:
+
+* Preventing unintended modification of module and class level
+  constants and documenting them as constants in a checkable way.
+* Creating a read-only attribute that may not be overridden by
+  subclasses. (``@property`` can make an attribute read-only but
+  does not prevent overriding)
+* Allowing a name to be used in situations where ordinarily a literal
+  is expected (for example as a file name for ``NamedTuple``, a tuple
```
Comment (JelleZijlstra) [inline pep-0591.rst]: Sorry I'm late to the party, but I'm guessing "file name" is a typo for "field name".

File: pep-0591.rst [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+
+   class Base:
+       @overload
+       def method(self) -> None: ...
+       @overload
+       def method(self, arg: int) -> int: ...
+       @final
+       def method(self, x=None):
+           ...
+
```
Comment (JelleZijlstra) [inline pep-0591.rst]: Perhaps say that use of final on a non-method function is an error.

File: pep-0591.rst [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+   class Base:
+       DEFAULT_ID: Final = 0
+
+   RATE = 300  # Error: can't assign to final attribute
+   Base.DEFAULT_ID = 1  # Error: can't override a final attribute
+
+   for x in [1, 2, 3]:
+      FOO: Final = x  # Error: Cannot use Final inside a loop
+
+error: Cannot use Final inside a loop
```
Comment (JelleZijlstra) [inline pep-0591.rst]: Stray line

**Comment by sproshev** [comment]
Is it possible to use `Final` in type comments?

**Comment by ilevkivskyi** [comment]
Yes.

**Comment by gvanrossum** [comment]
Maybe there should be an example of that somewhere? (At least in the PEP? And even Python 3 code might want to use type comments if it's straddling with 2.7.)

**Comment by msullivan** [comment]
I'll make a PR soon

**Comment by sproshev** [comment]
In the following sample mypy does not raise an error for `y`:

```
from typing_extensions import Final

x: Final[int] = 1
x: Final[str] = "str"

def foo():
    y: Final[int] = 1
    y: Final[str] = "str"
```

Is it a mypy bug or this case is allowed by PEP?

**Comment by ilevkivskyi** [comment]
@sproshev How about adding an annotation to `foo()` or using `--check-untyped-defs`?

Also please don't spam in this PR. If you have questions, ask them on typing Gitter.

**Pull request merged**